oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/IntrospectedOAuth2TokenAuthenticationToken.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.resource.authentication;
+
+import java.util.Collection;
+import java.util.Map;
+
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.SpringSecurityCoreVersion;
+import org.springframework.security.oauth2.core.OAuth2TokenWithClaims;
+
+/**
+ * An {@link org.springframework.security.core.Authentication} token that represents a successful authentication as
+ * obtained through an opaque token
+ * <a target="_blank" href="https://tools.ietf.org/html/rfc7662">introspection</a>
+ * process.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ */
+public class IntrospectedOAuth2TokenAuthenticationToken extends AbstractOAuth2TokenAuthenticationToken<OAuth2TokenWithClaims> {
+	private static final long serialVersionUID = SpringSecurityCoreVersion.SERIAL_VERSION_UID;
+
+	public IntrospectedOAuth2TokenAuthenticationToken(OAuth2TokenWithClaims token) {
+		super(token);
+	}
+
+	public IntrospectedOAuth2TokenAuthenticationToken(OAuth2TokenWithClaims token,
+			Collection<? extends GrantedAuthority> authorities) {
+
+		super(token, authorities);
+		setAuthenticated(true);
+	}
+
+	@Override
+	public Map<String, Object> getTokenAttributes() {
```

Comment:
It is strange that the token attributes are getting overridden to be the claims.

Reply:
Do you see it as inconsistent with the same behavior in [`JwtAuthenticationToken`](https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationToken.java#L64)? That is, they appear to be similar situations to me.

Is there a semantic difference between attributes and claims, e.g. attributes == verified claims? If so, then it would seem that `JwtAuthenticationToken` might be taking the wrong approach as well.

Reply:
I'm just wondering if the claims should be translated into attributes and injected as such.